src/macaron/slsa_analyzer/provenance/loader.py

```diff
@@ -103,9 +130,90 @@ def _load_provenance_file_content(
     if not isinstance(json_payload, dict):
         raise LoadIntotoAttestationError("The provenance payload is not a JSON object.")
 
+    if json_payload["predicate"]:
+        return json_payload
+
+    # For provenance without a predicate (e.g. PyPI), try to use the provenance certificate instead.
+    raw_certificate = json_extract(provenance, ["verification_material", "certificate"], str)
+    if not raw_certificate:
+        raise LoadIntotoAttestationError("Failed to extract certificate data.")
+    try:
+        decoded_certificate = base64.b64decode(raw_certificate)
+        certificate_claims = get_x509_certificate_values(decoded_certificate)
+    except UnicodeDecodeError as error:
+        raise LoadIntotoAttestationError("Cannot decode the payload.") from error
+    except ValueError as error:
+        logger.debug(error)
+        raise LoadIntotoAttestationError("Error parsing certificate.") from error
+
+    pypi_predicate = PyPICertificatePredicate.build_predicate(
+        certificate_claims["source_repo"],
+        certificate_claims["source_digest"],
+        certificate_claims["workflow"],
+        certificate_claims["invocation"],
+    )
+    json_payload["predicate"] = pypi_predicate
     return json_payload
 
 
+def get_x509_certificate_values(decoded_certificate: bytes) -> dict:
+    """Retrieve the values of interest from an x509 certificate.
+
+    Parameters
+    ----------
+    decoded_certificate: bytes
+        The decoded certificate bytes.
+
+    Returns
+    -------
+    dict
+        A dictionary of the extracted values.
+
+    Raises
+    ------
+    ValueError
+        If the values could not be extracted.
+    """
+    certificate = x509.load_der_x509_certificate(decoded_certificate)
+    certificate_claims = {}
+    for extension in certificate.extensions:
+        if extension.oid.dotted_string not in _OID_IDS:
+            continue
+
+        claim_name = _OID_IDS[extension.oid.dotted_string]
+        certificate_claims[claim_name] = extension.value.value
```

Comment:
Are we assuming that `extension` here is always of type [cryptography.x509.UnrecognizedExtension](https://cryptography.io/en/latest/x509/reference/#cryptography.x509.UnrecognizedExtension) because of the OIDs we are looking for are not part of the standard X.509 extension type. 
If true, I think it's good to have a small comment about it.